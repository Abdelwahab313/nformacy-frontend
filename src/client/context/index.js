import React, { createContext, useContext, useReducer } from 'react';
import {
  SET_ERROR_MESSAGE,
  UPDATE_CLIENTS,
  UPDATE_CURRENT_CLIENT,
  VERIFY_CLIENT,
  VERIFY_CURRENT_CLIENT,
} from './actionTypes';

const ClientContext = createContext();
const defaultValues = {
  clients: [],
  errorMessage: '',
};

const ClientProvider = ({ children, initialValue }) => {
  const { Provider } = ClientContext;
  initialValue = { ...defaultValues, ...initialValue };
  return (
    <Provider value={useReducer(clientReducer, initialValue)}>
      {children}
    </Provider>
  );
};

const useClientState = () => {
  const context = useContext(ClientContext);
  if (context === undefined) {
    throw new Error('useClientState must be used within a ClientProvider');
  }
  return context;
};

const clientReducer = (state, action) => {
  switch (action.type) {
    case UPDATE_CLIENTS:
      return {
        ...state,
        clients: action.payload,
      };
    case UPDATE_CURRENT_CLIENT:
      return {
        ...state,
        currentClient: action.payload,
      };
    case VERIFY_CURRENT_CLIENT:
      const currentClient = state.currentClient;
      currentClient.verified = true;
      return {
        ...state,
        currentClient: currentClient,
      };
    case VERIFY_CLIENT:
      const newClients = state.clients.map((client) => {
        if (client[0] === action.payload.uuid) {
          client[6] = true;
          return client;
        } else {
          return client;
        }
      });
      return {
        ...state,
        clients: newClients,
      };
    case SET_ERROR_MESSAGE:
      return {
        ...state,
        errorMessage: action.payload,
      };
  }
};

export { ClientProvider, useClientState };
