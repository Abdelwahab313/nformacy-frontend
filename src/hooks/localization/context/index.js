import React, { createContext, useContext, useReducer } from 'react';

const LocaleContext = createContext();
const DEFAULT_LOCALE = 'en';
const LOCALES = {
  en: 'en',
  ar: 'ar',
};
export const CHANGE_LOCALE = 'CHANGE_LOCALE';

export const useLocaleContext = () => {
  const context = useContext(LocaleContext);
  if (context === undefined)
    throw new Error('useLocale should only be used within locale provider');
  return context;
};

export const LocaleProvider = ({
  children,
  initialLocale = DEFAULT_LOCALE,
}) => {
  const { Provider } = LocaleContext;
  return (
    <Provider value={useReducer(localeReducer, initialLocale)}>{children}</Provider>
  );
};

const localeReducer = (state, action) => {
  if (action.type === CHANGE_LOCALE) {
    return action.payload.locale;
  }
  return LOCALES.en;
};
