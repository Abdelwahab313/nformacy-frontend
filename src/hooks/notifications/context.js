import React, { useContext, useReducer } from 'react';
import { ActionCableProvider } from '../useActionCable';
import { CHANNEL_URL } from '../../settings';
import { produce, current } from 'immer';
import { toast as showToast } from 'react-toastify';

const defaultStates = {
  notifications: [],
  menuOpened: null,
  unread: false,
  unreadCount: 0,
};

const NotificationsContext = React.createContext();

export const notificationActions = {
  notificationReceived: 'notificationReceived',
  menuToggled: 'menuToggled',
  menuClosed: 'menuClosed',
};

const MAX_NOTIFICATIONS = 10;
export const useNotificationsContext = () => {
  const context = useContext(NotificationsContext);
  if (context === undefined) {
    throw new Error(
      'useNotificationsContext must be used within a NotificationsProvider',
    );
  }
  return context;
};
const initContext = (initialValue) => {
  return initialValue || defaultStates;
};

export const NotificationsProvider = ({
  children,
  initialNotifications = [],
  unreadCount = 0,
}) => {
  const { Provider } = NotificationsContext;
  const initialState = {
    ...defaultStates,
    notifications: initialNotifications,
    unread: unreadCount > 0,
    unreadCount,
  };
  return (
    <Provider
      value={useReducer(NotificationsReducer, initialState, initContext)}>
      <ActionCableProvider url={CHANNEL_URL}>{children}</ActionCableProvider>
    </Provider>
  );
};

function Notification(notification) {
  return {
    notificationId: notification.notification_id,
    targetId: notification.target_id,
    messageKey: notification.message_key,
    createdAt: notification.created_at,
    type: notification.type,
    readAt: null,
  };
}

const updateUserInLocalStorage = (currentUser, updatedFields) => {
  localStorage.setItem(
    'user',
    JSON.stringify({ ...currentUser, ...updatedFields }),
  );
};

const NotificationsReducer = (state, action) => {
  switch (action.type) {
    case notificationActions.notificationReceived:
      const currentUser = action.payload.currentUser;
      const receivedNotification = Notification(action.payload.notification);
      showToast(receivedNotification.messageKey, {
        toastId: receivedNotification.notificationId,
      });
      return produce(state, (draftState) => {
        if (draftState.notifications.length === MAX_NOTIFICATIONS) {
          draftState.notifications.pop();
        }
        draftState.notifications.unshift(receivedNotification);
        draftState.unread = true;
        draftState.unreadCount = state.unreadCount + 1;
        updateUserInLocalStorage(currentUser, {
          unreadNotifications: draftState.unreadCount,
          notifications: current(draftState.notifications),
        });
      });
    case notificationActions.menuToggled:
      return produce(state, (draftState) => {
        if (state.menuOpened?.contains(action.payload.target)) {
          draftState.menuOpened = null;
        } else {
          draftState.menuOpened = action.payload.currentTarget;
          draftState.unread = false;
        }
      });
    case notificationActions.menuClosed:
      return produce(state, (draftState) => {
        draftState.menuOpened = null;
      });
    default:
      throw Error(
        `Action: ${action.type} is not defined in notifications reducer`,
      );
  }
};
