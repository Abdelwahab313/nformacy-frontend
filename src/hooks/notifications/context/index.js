import React, { useContext, useReducer } from 'react';
import { ActionCableProvider } from '../useActionCable';
import { CHANNEL_URL } from '../../../settings';
import {
  clearToast,
  closeMenu,
  loadNotifications,
  receiveNotification,
  toggleMenu,
  visitNotification,
} from './actions';

const defaultStates = {
  notifications: [],
  menuOpened: null,
  unread: false,
  unreadCount: 0,
};

const NotificationsContext = React.createContext();

export const notificationActions = {
  notificationsLoaded: 'notificationsLoaded',
  notificationReceived: 'notificationReceived',
  notificationVisited: 'notificationVisited',
  menuToggled: 'menuToggled',
  menuClosed: 'menuClosed',
  toastCleared: 'toastCleared',
};

export const useNotificationsContext = () => {
  const context = useContext(NotificationsContext);
  if (context === undefined) {
    throw new Error(
      'useNotificationsContext must be used within a NotificationsProvider',
    );
  }
  return context;
};
const initContext = (initialValue) => {
  return initialValue || defaultStates;
};

export const NotificationsProvider = ({ children }) => {
  const { Provider } = NotificationsContext;
  const initialState = {
    ...defaultStates,
    notifications: [],
    unreadCount: 0,
  };
  return (
    <Provider
      value={useReducer(NotificationsReducer, initialState, initContext)}>
      <ActionCableProvider url={CHANNEL_URL}>{children}</ActionCableProvider>
    </Provider>
  );
};

const NotificationsReducer = (state, action) => {
  switch (action.type) {
    case notificationActions.notificationsLoaded:
      return loadNotifications(state, action);
    case notificationActions.notificationReceived:
      return receiveNotification(action, state);
    case notificationActions.menuToggled:
      return toggleMenu(state, action);
    case notificationActions.menuClosed:
      return closeMenu(state);
    case notificationActions.notificationVisited:
      return visitNotification(state, action);
    case notificationActions.toastCleared:
      return clearToast(state);
    default:
      throw Error(
        `Action: ${action.type} is not defined in notifications reducer`,
      );
  }
};
