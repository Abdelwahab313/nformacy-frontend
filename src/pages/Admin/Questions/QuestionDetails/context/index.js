import React, { createContext, useContext, useReducer } from 'react';
import QuestionActionTypes from './actionTypes';

const defaultValues = {
  questionDetails: '',
};

const QuestionContext = createContext();

const QuestionProvider = ({ children, initialValue = {} }) => {
  initialValue = { ...defaultValues, ...initialValue };
  const { Provider } = QuestionContext;
  return (
    <Provider value={useReducer(QuestionReducer, initialValue)}>
      {children}
    </Provider>
  );
};

const useQuestionContext = () => {
  const context = useContext(QuestionContext);
  if (context === undefined) {
    throw new Error('useQuestionContext must be used within a AuthProvider');
  }
  return context;
};

const QuestionReducer = (state, action) => {
  switch (action.type) {
    case QuestionActionTypes.UPDATE_QUESTION:
      const updatedQuestion = { ...state.questionDetails, ...action.payload };
      return { ...state, questionDetails: updatedQuestion };

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

export { useQuestionContext, QuestionProvider };
