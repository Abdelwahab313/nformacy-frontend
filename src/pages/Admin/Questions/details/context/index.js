import React, { createContext, useContext, useReducer } from 'react';

const defaultValues = {
  questionDetails: '',
};

const QuestionContext = createContext();

const QuestionProvider = ({ children, initialValue = {} }) => {
  initialValue = { ...defaultValues, ...initialValue };
  const { Provider } = QuestionContext;
  return (
    <Provider value={useReducer(QuestionReducer, initialValue)}>
      {children}
    </Provider>
  );
};

const useQuestionContext = () => {
  const context = useContext(QuestionContext);
  if (context === undefined) {
    throw new Error('useQuestionContext must be used within a AuthProvider');
  }
  return context;
};

export const QuestionActionTypes = {
  UPDATE_QUESTION: 'UPDATE_QUESTION',
};

const QuestionReducer = (state, action) => {
  switch (action.type) {
    case QuestionActionTypes.UPDATE_QUESTION:
      return { ...state, questionDetails: action.payload };

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

export { useQuestionContext as useAuth, QuestionProvider as AuthProvider };
