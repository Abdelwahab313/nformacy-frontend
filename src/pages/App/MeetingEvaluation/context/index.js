import React, { createContext, useContext, useReducer } from 'react';
// import CallEvaluationActionTypes from './actionTypes';

const defaultValues = {
  comment: '',
  ratingEvaluations: {}
};

const CallEvaluationContext = createContext();

const CallEvaluationProvider = ({ children, initialValue = {} }) => {
  initialValue = { ...defaultValues, ...initialValue };
  const { Provider } = CallEvaluationContext;
  return (
    <Provider value={useReducer(CallEvaluationReducer, initialValue)}>
      {children}
    </Provider>
  );
};

const useCallEvaluationContext = () => {
  const context = useContext(CallEvaluationContext);
  if (context === undefined) {
    throw new Error('useCallEvaluationContext must be used within a CallEvaluationContext');
  }
  return context;
};

const CallEvaluationReducer = (state, action) => {
  switch (action.type) {

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};

export { useCallEvaluationContext, CallEvaluationProvider };
