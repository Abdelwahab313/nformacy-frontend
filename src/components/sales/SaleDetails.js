import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router';
import { makeStyles } from '@material-ui/core';
import { useAuth } from '../../context/auth';
import { getSaleWithUUID } from '../../apis/salesApi';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';
import WarningIcon from '@material-ui/core/SvgIcon/SvgIcon';
import { fetchClient } from '../../apis/clientsApi';
import { fetchUser } from '../../apis/usersApi';
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import { fetchProduct } from '../../product/productsApi';

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1,
    background: '#F5F5F5',
    padding: theme.spacing(1),
  },
  paper: {
    marginRight: theme.spacing(3),
    paddingLeft: theme.spacing(1),
    minWidth: '340px',
    textAlign: 'center',
    color: theme.palette.text.secondary,
    width: '100%',
  },
  emptyContainer: {
    position: 'absolute',
    left: '50%',
    top: '50%',
    transform: 'translate(-50%, -50%)',
  },
  productList: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    margin: theme.spacing(2),
  },
  notFound: {
    color: 'red',
  },
}));
const SaleDetails = (props) => {
  const classes = useStyles();
  const { uuid } = useParams();
  const { authTokens } = useAuth();
  const [saleData, setSaleData] = useState(null);
  const [errors, setErrors] = useState();
  const [saleLoading, setSaleLoading] = useState(true);
  const [notFound, setNotFound] = useState(false);

  useEffect(() => {
    let tempSaleData;
    getSaleWithUUID(uuid, authTokens)
      .then((response) => {
        tempSaleData = response.data;
      })
      .then(async () => {
        const response = await fetchClient(tempSaleData.to, authTokens);
        tempSaleData = { ...tempSaleData, to: response.data };
      })
      .then(async () => {
        const response = await fetchUser(tempSaleData.by, authTokens);
        tempSaleData = { ...tempSaleData, by: response.data };
      })
      .then(async () => {
        await Promise.all(
          tempSaleData.products.map(async (product) => {
            const response = await fetchProduct(product.uuid, authTokens);
            delete product.uuid;
            product.data = response.data;
          }),
        );
      })
      .then(() => {
        setSaleData({
          ...tempSaleData,
        });
      })
      .catch((error) => {
        setErrors(error);
        setNotFound(true);
      })
      .finally(() => {
        setSaleLoading(false);
      });
  }, [uuid]);
  if (saleLoading) {
    return (
      <div className={classes.emptyContainer}>
        <CircularProgress />
      </div>
    );
  } else if (notFound) {
    return (
      <div className={classes.emptyContainer}>
        <Typography variant='h3' className={classes.notFound} gutterBottom>
          الفاتورة المطلوبة غير موجودة{' '}
          <WarningIcon className={classes.largeIcon} />
        </Typography>
      </div>
    );
  }
  if (saleLoading) {
    return (
      <div className={classes.emptyContainer}>
        <CircularProgress />
      </div>
    );
  } else {
    return (
      <div className={classes.root} dir='rtl'>
        <div className={classes.productList}>
          <Paper className={classes.paper}>
            <p id={'title'}>تفاصيل الفاتورة</p>
            <Table>
              <TableBody>
                <TableRow>
                  <TableCell colSpan={2} align='right'>
                    التاريخ
                  </TableCell>
                  <TableCell colSpan={2} align='left'>
                    {(() => {
                      const date = new Date(saleData.created);
                      return `${date.toLocaleTimeString()} ${date.toLocaleDateString()}`;
                    })()}
                  </TableCell>
                </TableRow>
                <TableRow>
                  <TableCell>أسم العميل</TableCell>
                  <TableCell>{saleData.to.name}</TableCell>
                  <TableCell>أسم المندوب</TableCell>
                  <TableCell>{`${saleData.by.first_name} ${saleData.by.last_name}`}</TableCell>
                </TableRow>
              </TableBody>
            </Table>
          </Paper>
        </div>
        <div className={classes.productList}>
          <Paper className={classes.paper}>
            <p id={'title'}>تفاصيل المنتجات</p>
            <Table id={'client-info'}>
              <TableBody>
                <TableRow id={'tableHeader'}>
                  <TableCell>المنتج</TableCell>
                  <TableCell>السعر</TableCell>
                  <TableCell>الكمية</TableCell>
                  <TableCell>الاجمالى</TableCell>
                </TableRow>
                {saleData.products.map((product) => {
                  return (
                    <TableRow key={product.data.uuid} id={'clientName'}>
                      <TableCell>{product.data.name}</TableCell>
                      <TableCell>{product.price}</TableCell>
                      <TableCell>{product.quantity}</TableCell>
                      <TableCell>
                        {parseFloat(
                          product.price.substring(4, product.price.length),
                        ) * parseFloat(product.quantity)}
                      </TableCell>
                    </TableRow>
                  );
                })}
                <TableRow id={'address'}>
                  <TableCell colSpan={3} align='right'>
                    الحساب الكلي
                  </TableCell>
                  <TableCell colSpan={1} align='right'>
                    {saleData.total_price}
                  </TableCell>
                </TableRow>
              </TableBody>
            </Table>
          </Paper>
        </div>
      </div>
    );
  }
};
export default SaleDetails;
