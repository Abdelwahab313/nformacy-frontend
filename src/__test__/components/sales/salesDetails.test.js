import React from 'react';
import { shallow } from 'enzyme';
import * as AuthContextMoudle from '../../../context/auth';
import { AuthContext } from '../../../context/auth';
import SaleDetails from '../../../components/sales/SaleDetails';

describe('SaleDetails', () => {
  let tree;
  let wrapper;

  beforeEach(() => {
    const reactRouter = require('react-router');
    jest.spyOn(AuthContextMoudle, 'useAuth').mockImplementation(() => ({
      authTokens: '',
    }));
    jest.spyOn(reactRouter, 'useParams').mockImplementation(() => ({
      uuid: 'test-uuid',
    }));
    tree = (
      <AuthContext.Provider>
        <SaleDetails />
      </AuthContext.Provider>
    );
    wrapper = shallow(tree).dive();
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
