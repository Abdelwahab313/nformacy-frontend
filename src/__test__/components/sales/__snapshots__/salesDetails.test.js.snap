// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaleDetails should match snapshot 1`] = `ShallowWrapper {}`;
